lib/ansible/modules/source_control/git.py

```diff
@@ -977,6 +988,57 @@ def main():
     track_submodules = module.params['track_submodules']
 
     result.update(before=None)
+
+    if archive:
+        if module.check_mode:
+            result.update(changed=True)
+            module.exit_json(**result)
+
+        # Archive already exists, then fail
+        if os.path.exists(archive):
+            module.fail_json(msg="Archive file %s already exists" % archive,
+                             details="Please remove already created archive "
+                                     "file before proceeding")
+
+        all_archive_fmt = {'.zip': 'zip', '.gz': 'tar.gz', '.tar': 'tar'}
+        _, archive_ext = os.path.splitext(archive)
+        archive_fmt = all_archive_fmt.get(archive_ext, None)
+        if archive_fmt is None:
+            module.fail_json(msg="Unable to get file extension from "
+                                 "archive file name : %s" % archive,
+                             details="Please specify archive as filename with "
+                                     "extension. File extension can be one "
+                                     "of ['tar', 'tar.gz', 'zip']")
+
+        try:
+            os.makedirs(dest)
+        except OSError:
+            pass
+
+        repo_name = repo.split("/")[-1].replace(".git", "")
+        # Git archive is not supported by all git servers, so
+        # we will first clone and perform git archive from local directory
+        repo_dest = os.path.join(dest, repo_name)
+
+        if not os.path.exists(repo_dest):
+            cmd = "%s clone %s %s" % (git_path, repo, repo_name)
+            (rc, out, err) = module.run_command(cmd, cwd=dest)
+            if rc != 0:
+                module.fail_json(msg="Failed to clone source",
+                                 details="Git clone failed, try again "
+                                         "after removing %s" % repo_dest)
```

Comment:
Is this the only way this can fail? If not, might be nice to provide the out, err in the message, if this is the only way this can fail this is fine though.